### PR TITLE
✨ Improve progress and exception handling

### LINKED ISSUE #772: Handle sync sequence exceptions correctly
## Description
We haven't really implemented exception handling at all [in SyncSequence ](https://github.com/toggl/android/blob/e895cf71ee5c42f3fc41c7224b9d1e8c5eda53c2/sync/src/main/java/com/toggl/sync/sequence/SyncSequence.kt#L27)

[Xamarin code](https://github.com/toggl/mobileapp/blob/54bb2f1194b97fac59113222caaf390a4acf1581/Toggl.Core/Sync/SyncManager.cs#L166) for reference

## Dependencies
none

## Definition of done
- [ ] Handle sync exceptions properly in `SyncSequence`
- [ ] Or in `SyncWorker` if needed

### LINKED ISSUE #844: Add correct icon for sync notification
## Description
The icon we now use in "sync in progress" notification is completely random. We need a real one. We could simply use the standard [material sync icon](https://material.io/resources/icons/?icon=cached&style=baseline) but this needs to be decided in cooperation with designers - maybe they can think of something more uniquely Toggl.

The current provisional implementation looks like this:

**Status bar:**

Screen Shot 2020-09-11 at 12 34 59

**Expanded notification:**

Screen Shot 2020-09-11 at 12 34 59

## Dependencies
None

## Definition of done
- [x] The new icon now displayed in sync notification

### PR DESCRIPTION
**Pull request by semanticer**
### Summary
This PR implements basic exception handling in sync sequence. It also improves the debug listener and adds progress reporting listener.

### Technical considerations
~~This is not a complete equivalent of [xamarin code](https://github.com/toggl/mobileapp/blob/54bb2f1194b97fac59113222caaf390a4acf1581/Toggl.Core/Sync/SyncManager.cs#L166) yet, some of the exceptions from Xamarin are not even defined in Aurora so I'm not sure if we should create some further issues to take care of this or...?~~ 🤷 

### Relationships

Closes #772 - ~~kinda~~
Closes #844

### Pull Request checklist
- [x] The code does not hinder the app's performance
- [x] The code does not introduce regressions
- [ ] The code in this PR is tested thoroughly enough that changing the behavior to an incorrect one breaks tests
- [ ] One of our [live or file custom templates](https://github.com/toggl/android/tree/master/.idea/templates) was used in this PR

### Pointers
You can check the log if everything is reported correctly, you can also try `syncLauncher.enqueueSync().observe( ... )` to verify you are getting some progress data back. 

And you can also try throwing some exception in one of the sync steps and verify it has been caught correctly.

### Review template

- [ ] The code is thoroughly tested in well written and resilient manner
- [ ] The code is performant and leads to an optimal user experience
- [ ] The code is testable and does not cut corners for the sake of delivering
- [ ] The code considers older platforms well and will work in older devices just like it will in new ones

- [ ] The code considers light and dark theme
- [ ] The code considers RTL and does not mention left/right
- [ ] The code considers i18n
  - [ ] There are no hardcoded strings in the UI
  - [ ] All strings are included in every language we support OR there's an issue requesting new translations
- [ ] There are no hard-coded numbers referring to dimensions or font sizes
- [ ] The code considers UI elements that are similar and extracts their similarities into styles to avoid duplication

## Merge permissions
🦑

### CODE REVIEW

File: sync/src/main/java/com/toggl/sync/SyncLauncher.kt [hunk 1]
```diff
@@ -16,9 +19,9 @@ class SyncLauncher @Inject constructor(
     private val syncSequence: SyncSequence,
     private val workManagerProvider: WorkManagerProvider
 ) {
-    fun enqueueSync() {
+    fun enqueueSync(): LiveData<WorkInfo?>? {
```
Comment (heytherewill) [inline sync/src/main/java/com/toggl/sync/SyncLauncher.kt:22]: Please expose a Flow

File: sync/src/main/java/com/toggl/sync/SyncWorker.kt [hunk 2]
```diff
@@ -21,23 +22,17 @@ class SyncWorker @WorkerInject constructor(
...
-        override suspend fun onSequenceRunStarted(runCounter: Int) {
-            setForeground(createForegroundInfo("Run $runCounter started"))
-        }
-    }
-
-    private val allListeners = listOf(listener, debugSyncListener)
+    private val allListeners = listOf(
+        listener,
+        SyncProgressListener(syncSequence.stepCount, this::setProgress),
+        SyncDebugListener(syncSequence.stepCount) { msg -> setForeground(createForegroundInfo(msg)) }
```
Comment (heytherewill) [inline sync/src/main/java/com/toggl/sync/SyncWorker.kt:28]: Why not inject all of these using `@Binds @IntoSet`?

File: sync/src/main/java/com/toggl/sync/SyncWorker.kt [hunk 3]
```diff
@@ -48,7 +43,7 @@ class SyncWorker @WorkerInject constructor(
             .setContentTitle(title)
             .setTicker(title)
             .setContentText(progress)
-            .setSmallIcon(android.R.drawable.ic_popup_reminder)
+            .setSmallIcon(android.R.drawable.ic_popup_reminder) // TODO we need a better icon
```
Comment (heytherewill) [inline sync/src/main/java/com/toggl/sync/SyncWorker.kt:46]: Please ping designers or create an issue!

**Review by heytherewill — CHANGES_REQUESTED** [review]

File: sync/src/main/java/com/toggl/sync/SyncWorker.kt (hunk 3 above)
Comment (semanticer) [reply sync/src/main/java/com/toggl/sync/SyncWorker.kt:46]: https://github.com/toggl/android/issues/844

File: sync/src/main/java/com/toggl/sync/sequence/SyncSequence.kt [hunk 4]
```diff
@@ -72,6 +73,8 @@ interface SyncStep {
     suspend fun run()
 }
 
+typealias SyncSequenceListenerProvider = (ProgressReporter) -> SyncSequenceListener
```
[the remark this answers is not in the document]
Comment (semanticer) [reply sync/src/main/java/com/toggl/sync/sequence/SyncSequence.kt]: I am sooooo Kotlin 1.3

**Comment by simonrozsival** [comment]
> This is not a complete equivalent of xamarin code yet, some of the exceptions from Xamarin are not even defined in Aurora so I'm not sure if we should create some further issues to take care of this or...? 🤷

I guess this PR doesn't close #772 - we really need to handle all those cases (`NoDefaultWorkspaceException`, `ClientDeprecatedException`, `ApiDeprecatedException` (teapoting), `UnauthorizedException` (API token was reset - soft logout)). I think you can merge this PR as long as the issue #772 stays open.

**Review by simonrozsival — APPROVED** [review]
All the exceptions are there now 👍

**Review by heytherewill — APPROVED** [review]
Looks great! One thing that I want you to change: Please change the `SyncStep` enum that is currently part of the AppState into the nicer `SyncProgress` sealed class for consistency 💪 

I'm approving for :sanic: though

**Pull request merged**